fio-plot: 1.1.16

johnrichardrinehart opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Sorry ahead-of-time if this PR is absolute crap. I have little to no experience with `python` packaging and I have attention issues, so there may be a lot of small things to clean up, here. Thanks for your review.

The goal of this PR is to add `fio-plot` (a package which provides two binaries, `fio-plot` and `bench_fio`), to `legacyPackages`. I've only tested that the binaries build and seem to run on `x86_64-linux`:
```bash
$ ./result/bin/bench_fio                                                                                                                
　
usage: bench_fio [-h] -d TARGET [TARGET ...] -t {device,file,directory,rbd} [-P CEPH_POOL] [-s SIZE] -o OUTPUT [-b BLOCK_SIZE [BLOCK_SIZE ...]]
                 [--iodepth IODEPTH [IODEPTH ...]] [--numjobs NUMJOBS [NUMJOBS ...]] [--runtime RUNTIME] [-p] [--precondition-repeat]
                 [--precondition-template PRECONDITION_TEMPLATE] [-m MODE [MODE ...]] [--rwmixread RWMIXREAD [RWMIXREAD ...]] [-e ENGINE] [--direct DIRECT] [--loops LOOPS]
                 [--time-based] [--entire-device] [--ss SS] [--ss-dur SS_DUR] [--ss-ramp SS_RAMP] [--extra-opts EXTRA_OPTS [EXTRA_OPTS ...]] [--invalidate INVALIDATE]
                 [--quiet] [--loginterval LOGINTERVAL] [--dry-run] [--destructive] [--remote REMOTE] [--remote-checks] [--remote-timeout REMOTE_TIMEOUT] [--create]
                 [--parallel]
bench_fio: error: the following arguments are required: -d/--target, -t/--type, -o/--output
$ ./result/bin/fio-plot 
usage: fio-plot [-h] -i INPUT_DIRECTORY [INPUT_DIRECTORY ...] [-o OUTPUT_FILENAME] -T TITLE [-s SOURCE] (-L | -l | -N | -H | -g | -C) [--disable-grid] [--enable-markers]
                [--subtitle SUBTITLE] [-d IODEPTH [IODEPTH ...]] [-n NUMJOBS [NUMJOBS ...]] [-M [MAXDEPTH]] [-J [MAXJOBS]] [-D [DPI]] [-p [PERCENTILE]]
                -r {read,write,randread,randwrite,randrw,trim,rw,readwrite,randtrim,trimwrite} [-m MAX_Z] [-e MOVING_AVERAGE] [--min-iops MIN_IOPS] [--min-lat MIN_LAT]
                [-t {bw,iops,lat,slat,clat} [{bw,iops,lat,slat,clat} ...]] [-f {read,write} [{read,write} ...]] [--truncate-xaxis TRUNCATE_XAXIS]
                [--xlabel-depth XLABEL_DEPTH] [--xlabel-parent XLABEL_PARENT] [--xlabel-segment-size XLABEL_SEGMENT_SIZE] [--xlabel-single-column] [-w LINE_WIDTH]
                [--group-bars] [--show-cpu] [--show-data] [--show-ss] [--table-lines] [--max-lat MAX_LAT] [--max-clat MAX_CLAT] [--max-slat MAX_SLAT] [--max-iops MAX_IOPS]
                [--max-bw MAX_BW] [--draw-total] [--colors COLORS [COLORS ...]] [--disable-fio-version] [--title-fontsize TITLE_FONTSIZE]
                [--subtitle-fontsize SUBTITLE_FONTSIZE] [--source-fontsize SOURCE_FONTSIZE] [--credit-fontsize CREDIT_FONTSIZE] [--table-fontsize TABLE_FONTSIZE]
                [--include-hosts INCLUDE_HOSTS [INCLUDE_HOSTS ...] | --exclude-hosts EXCLUDE_HOSTS [EXCLUDE_HOSTS ...]]
fio-plot: error: the following arguments are required: -i/--input-directory, -T/--title, -r/--rw
```

Now, in order to provide the application package, I also needed to add the corresponding `python` module package, `fio_plot`. But, `fio_plot` depends on https://github.com/Technologicat/pyan/, which is not in tree, yet.

Also, I was running into weird issues around `bin` files already existing when the `fio_plot` repository _only_ provided its `setup.py`. The world seems to be moving toward `pyproject.toml` so you might notice that I've also bundled in a patch to the source repository which provides a `pyproject.toml`. This file unblocks the build. I'm going to make a corresponding PR to https://github.com/louwrentius/fio-plot/, but I'm not sure if this project is still maintained. I also want to be able to point to this PR in that PR and say "hey, I needed to hack something up in `nixpkgs` because your project packaging is stale". I'm happy to remove this patch later if the source repository includes these changes.

So, rather then break this into 3 separate/stacked PRs, I decided to provide a single PR. I hope that's okay.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossSmyth reviewed (changes requested):
Mostly good, a few little changes:

1. Some sanity checks
2. Remove `with lib;`
3. Use `python3Packages` rather than `python3.pkgs` because the latter doesn't splice correctly
4. Is there an upstream PR for the pyproject.toml? I would recommend putting your patch there if not, then fetching it.

johnrichardrinehart commented:
@RossSmyth Thanks for the deep review. I've:
1. created a PR to https://github.com/louwrentius/fio-plot/pull/156 and used that commit to pull in my patch and removed my in-tree patch file;
2. removed `with lib;` in the various `meta` sections;
3. added a `pythonImportsCheck` for `pyan3` (module names itself "pyan", apparently);
4. removed things like `python3Packages.foo` and instead moved `foo` into the args directly;
5. made sure that it still builds;

I think that's everything. Please let me know if there's more work to do!

RossSmyth reviewed (changes requested):
I didn't notice this when I was reviewing earlier. Since it is just a different view of the same source you can just have this.

RossSmyth commented:
You also need to squash your commits to follow the commit convention.

johnrichardrinehart commented:
> You also need to squash your commits to follow the commit convention.

Alright. I'll do that if the current state of all the files look good (as of 4558e93e7d269e3750083f03c659637847395294). Let me know and then I'll squash. I kept the history around in case I needed to revert and to support the review.

RossSmyth commented on pkgs/by-name/fi/fio-plot/package.nix:
```diff
@@ -0,0 +1,6 @@
+  python3Packages,
+}:
+(python3Packages.toPythonApplication python3Packages.fio_plot).overrideAttrs (old: {
+  meta.mainProgram = "fio-plot";
```
You can check the output with `nix eval -f ./. fio-plot.meta`
```suggestion
  meta.mainProgram = old.meta // { mainProgram = "fio-plot"; };
```

johnrichardrinehart replied:
Ah, nice. Changed!

RossSmyth reviewed (changes requested):
Last thing

RossSmyth commented:
Oops, for the commit structure it should be three commits:
1. python3Packages.pyan3: init at v1.2.0
2. python3Packages.fio_plot: init at 1.1.16
3. fio-plot: init at 1.1.16

One for each new package added.

johnrichardrinehart committed:
python3Packages.pyan3: init at v1.2.0

johnrichardrinehart committed:
python3Packages.fio_plot: init at 1.1.16

fix: fio_plot tests

chore: abbreviate pyan3 description

johnrichardrinehart committed:
fio-plot: init at 1.1.16

johnrichardrinehart commented:
> 3\. fio-plot: init at 1.1.16

Okay. I think we're stable.

johnrichardrinehart commented:
> Oops, for the commit structure it should be three commits:
> 
> 1. python3Packages.pyan3: init at v1.2.0
> 2. python3Packages.fio_plot: init at 1.1.16
> 3. fio-plot: init at 1.1.16
> 
> One for each new package added.

Okay, I think we're stable!

RossSmyth reviewed (approved):
Nice!

RossSmyth commented on pkgs/development/python-modules/pyan3/default.nix:
```diff
@@ -0,0 +1,36 @@
+    homepage = "https://github.com/Technologicat/pyan";
+    license = lib.licenses.gpl2;
+    maintainers = with lib.maintainers; [ johnrichardrinehart ];
+  };
```
This is a package. It does not have a mainProgram.
